### PR TITLE
Fix performance issue related to the rootNodes array

### PR DESCRIPTION
**Pull request by barroij**
All the nodes that does not have any parent are store in an array since commit  9094b709 (part of https://github.com/BabylonJS/Babylon.js/issues/5050). 

This change caused a performance regression as each time a node is switched from a "no parent" state to a "with parent" state, ther is a O(n) lookup in the array in order to remove it.  And it happens even if a parent is passed in the constructor of a Mesh, as the the addition to the array is always done in the Node base class constructor.

This PR fix this issue by storing and keeping up-to-date, in the Node, the index within the rootNodes array, so that removal can be done in O(1)

Please note that it assumes that the rootNodes array is NEVER modified from elsewhere. If so, it is up to the culprit to make sure that the indices remain up-to-date

Some discussion about this issue and the profiling that leads to this PR can be found here : http://www.html5gamedevs.com/topic/40058-linesmesh-and-createinstance/?do=findComment&comment=228801

### CODE REVIEW

File: src/Mesh/babylon.mesh.ts [hunk 1]
```diff
@@ -281,9 +281,6 @@
                     Tags.AddTagsTo(this, Tags.GetTags(source, true));
                 }
 
-                // Parent
-                this.parent = source.parent;
```
Comment (barroij) [inline src/Mesh/babylon.mesh.ts]: useless, it is set by the `Tools.DeepCopy`above :
```javascript
                 Tools.DeepCopy(source, this, ["name", "material", "skeleton", "instances", "parent", "uniqueId",
                     "source", "metadata", "hasLODLevels", "geometry", "isBlocked", "areNormalsFrozen",
                     "source", "metadata", "hasLODLevels", "geometry", "isBlocked", "areNormalsFrozen",
                     "onBeforeDrawObservable", "onBeforeRenderObservable", "onAfterRenderObservable", "onBeforeDraw"
                     "onBeforeDrawObservable", "onBeforeRenderObservable", "onAfterRenderObservable", "onBeforeDraw"
                 ],
]
````

Comment (deltakosh) [reply src/Mesh/babylon.mesh.ts]: source is in the avoid list

File: src/babylon.node.ts [hunk 2]
```diff
@@ -110,6 +110,9 @@
         /** @hidden */
         public _worldMatrixDeterminant = 0;        
 
+        /** @hidden */
+        private _sceneRootNodsIndex = -1;
```
Comment (deltakosh) [inline src/babylon.node.ts]: NodEs

File: src/babylon.node.ts [hunk 3]
```diff
@@ -135,10 +138,7 @@
                     this._parentNode._children.splice(index, 1);
                 }
 
-                if (!parent) {
-                    // Need to add this node to the rootNodes
-                    this._scene.rootNodes.push(this);
-                }
+                if (!parent) this.addToSceneRootNodes();
```
Comment (deltakosh) [inline src/babylon.node.ts]: can you add {} and a line feed? (just cosmetic)

File: src/babylon.node.ts [hunk 4]
```diff
@@ -151,14 +151,7 @@
...
 
-                if (!previousParentNode) {
-                    // Need to remove from rootNodes
-                    const rootNodeIndex = this._scene.rootNodes.indexOf(this);
-
-                    if (rootNodeIndex > -1) {
-                        this._scene.rootNodes.splice(rootNodeIndex, 1);
-                    }
-                }
+                if (!previousParentNode) this.removeFromSceneRootNodes();
```
Comment (deltakosh) [inline src/babylon.node.ts]: can you add {} and a line feed? (just cosmetic)

Comment (Hersir88) [reply src/babylon.node.ts]: would be nice in future to add tslint rules for such cases

File: src/Mesh/babylon.mesh.ts (hunk 1 above)
Comment (barroij) [reply src/Mesh/babylon.mesh.ts]: will remove, not related to the PR anyway

**Comment by barroij** [comment]
@deltakosh should be better now.

File: src/babylon.node.ts [hunk 5]
```diff
@@ -110,6 +110,9 @@
         /** @hidden */
         public _worldMatrixDeterminant = 0;        
 
+        /** @hidden */
+        private _sceneRootNodesIndex = -1;
+
```
Comment (barroij) [inline src/babylon.node.ts]: Note that if the order of children of a node does not matter, we could also keep the `_parentChildNodeIdx`to get a O(1) time removal from parent children array

File: src/babylon.node.ts (hunk 4 above)
Comment (barroij) [reply src/babylon.node.ts]: I agree, everything that could ensure coherent and automatic formatting is worth beeing added (tslin, prettier, pre-commit hook).

Another example, all the trailling whitespace should be removed they just add noise if your editor displays them and are likely to cause useless change in commit.

Comment (deltakosh) [reply src/babylon.node.ts]: I agree. If someone wants to volunteer to set it up in the project, I'll appreciate the help

File: src/babylon.node.ts (hunk 5 above)
Comment (deltakosh) [reply src/babylon.node.ts]: Yes agree

**Pull request merged**